### PR TITLE
Add LNET statistics

### LINKED ISSUE #70: Add lnet statistics
Use this issue to track our effort to add lnet statistics.

### PR DESCRIPTION
**Pull request by roclark**
Fixes #70 

Signed-Off-By: Robert Clark <robert.d.clark@hpe.com>

### CODE REVIEW

**Comment by roclark** [comment]
@joehandzik this is obviously a huge addition, so let me know if you want me to break this up at all, though to me, it seemed best to keep them all in the same PR.

File: sources/procfs.go [hunk 1]
```diff
@@ -123,24 +97,12 @@ func init() {
 	Factories["procfs"] = newLustreSource
 }
 
-type lustreSource struct {
+type lustreProcFSSource struct {
```
Comment (joehandzik) [inline sources/procfs.go]: Probably should name this 'lustreProcfsSource

File: sources/procsys.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	lnetReceiveLengthHelp string = "Total number of bytes received"
+	lnetRouteLengthHelp   string = "Total number of bytes for routed messages"
+	lnetDropLengthHelp    string = "Total number of bytes that have been dropped"
+)
+
+func init() {
+	Factories["procsys"] = newLustreProcSysSource
+}
+
+type lustreProcSysSource struct {
```
Comment (joehandzik) [inline sources/procsys.go]: And, similarly, 'lustreProcsysSource'

File: sources/procsys.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+func (s *lustreProcSysSource) generateLNETTemplates() error {
+	metricMap := map[string][]lustreHelpStruct{
+		"lnet": {
+			{"catastrophe", "catastrophe_enabled", "Returns 1 if currently in catastrophe mode", s.gaugeMetric},
+			{"console_backoff", "console_backoff_enabled", "Returns non-zero number if console_backoff is enabled", s.gaugeMetric},
+			{"console_max_delay_centisecs", "console_max_delay_centiseconds", "Minimum time in centiseconds before the console logs a message", s.gaugeMetric},
+			{"console_min_delay_centisecs", "console_min_delay_centiseconds", "Maximum time in centiseconds before the console logs a message", s.gaugeMetric},
+			{"console_ratelimit", "console_ratelimit_enabled", "Returns 1 if the console message rate limiting is enabled", s.gaugeMetric},
+			{"debug_mb", "debug_megabytes", "Maximum buffer size in megabytes for the LNET debug messages", s.gaugeMetric},
+			{"fail_err", "fail_err", "Number of errors that have been thrown", s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: fail_err_total or fail_error_total

File: sources/procsys.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+			{"fail_val", "fail_val", "Maximum number of times to fail", s.gaugeMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: fail_val_total or fail_value_total

File: sources/procsys.go (hunk 3 above)
Comment (joehandzik) [inline sources/procsys.go]: maybe even just 'error_total'.

File: sources/procsys.go (hunk 4 above)
Comment (joehandzik) [inline sources/procsys.go]: This is a weird one. Maybe 'error_maximum' or 'fail_maximum'

File: sources/procsys.go (hunk 3 above)
Comment (joehandzik) [inline sources/procsys.go]: 'fail_total' is on the table too. Geez these two metrics are weird.

File: sources/procsys.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+			{"lnet_memused", "lnet_memory_used_bytes", "Number of bytes allocated by LNET", s.gaugeMetric},
+			{"panic_on_lbug", "panic_on_lbug_enabled", "Returns 1 if panic_on_lbug is enabled", s.gaugeMetric},
+			{"stats", "allocated", lnetAllocatedHelp, s.gaugeMetric},
+			{"stats", "maximum", lnetMaximumHelp, s.counterMetric},
+			{"stats", "errors", lnetErrorsHelp, s.counterMetric},
+			{"stats", "send_count", lnetSendCountHelp, s.counterMetric},
+			{"stats", "receive_count", lnetReceiveCountHelp, s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: receive_count_total

File: sources/procsys.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+			{"console_ratelimit", "console_ratelimit_enabled", "Returns 1 if the console message rate limiting is enabled", s.gaugeMetric},
+			{"debug_mb", "debug_megabytes", "Maximum buffer size in megabytes for the LNET debug messages", s.gaugeMetric},
+			{"fail_err", "fail_err", "Number of errors that have been thrown", s.counterMetric},
+			{"fail_val", "fail_val", "Maximum number of times to fail", s.gaugeMetric},
+			{"lnet_memused", "lnet_memory_used_bytes", "Number of bytes allocated by LNET", s.gaugeMetric},
+			{"panic_on_lbug", "panic_on_lbug_enabled", "Returns 1 if panic_on_lbug is enabled", s.gaugeMetric},
+			{"stats", "allocated", lnetAllocatedHelp, s.gaugeMetric},
+			{"stats", "maximum", lnetMaximumHelp, s.counterMetric},
+			{"stats", "errors", lnetErrorsHelp, s.counterMetric},
+			{"stats", "send_count", lnetSendCountHelp, s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: send_count_total

File: sources/procsys.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+			{"stats", "receive_count", lnetReceiveCountHelp, s.counterMetric},
+			{"stats", "route_count", lnetRouteCountHelp, s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: route_count_total

File: sources/procsys.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+			{"stats", "drop_count", lnetDropCountHelp, s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: drop_count_total

File: sources/procsys.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+			{"stats", "send_length", lnetSendLengthHelp, s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: send_count_bytes_total

File: sources/procsys.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+			{"stats", "receive_length", lnetReceiveLengthHelp, s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: receive_bytes_total

File: sources/procsys.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+			{"stats", "route_length", lnetRouteLengthHelp, s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: route_bytes_total

File: sources/procsys.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+			{"stats", "drop_length", lnetDropLengthHelp, s.counterMetric},
```
Comment (joehandzik) [inline sources/procsys.go]: drop_bytes_total

File: sources/procsys.go (hunk 3 above)
Comment (roclark) [inline sources/procsys.go]: Decided with `fail_error_total` as it seemed to be the most accurate... though I agree about these metrics being weird.

**Comment by joehandzik** [comment]
@roclark This looks good to me. Do you want to wait on it until tomorrow so @knweiss has a chance to chime in, or should I merge now?

**Comment by roclark** [comment]
No real rush on my end, so maybe wait until tomorrow and merge if we don't hear anything.

**Pull request merged**